### PR TITLE
Did you mean

### LINKED ISSUE #177: Improve Did you mean
#163 kind of works but is still sometimes wonky.

Probably best idea is to have separate file for this. Mixing it with documentation has some downsides.

### PR DESCRIPTION
**Pull request by 5K1PY**
Fixes #177

### CODE REVIEW

File: pisek/config/task_config.py [hunk 1]
```diff
@@ -236,6 +239,18 @@ def validate_model(self):
                 {"task_type": self.task_type, "out_check": self.out_check},
             )
 
+        JUDGE_TYPES = {
+            TaskType.batch: [None, JudgeType.opendata_v1, JudgeType.cms_batch],
```
Comment (SvizelPritula) [inline pisek/config/task_config.py:243]: Isn't there a new format?

File: pisek/config/did_you_mean.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    from .config_hierarchy import ConfigHierarchy
+
+CONFIG_DESCRIPTION = str(files("pisek").joinpath("config/config-description"))
+
+
+def regex_cmp(regex: str, name: str) -> int:
+    if re.match(regex, name):
+        return 0
+    else:
+        return 5
```
Comment (SvizelPritula) [inline pisek/config/did_you_mean.py:21]: It's not clear why a function called `regex_cmp` returns a random int.

File: pisek/config/did_you_mean.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    def check(self, section: str, config: "ConfigHierarchy") -> str:
+        section = self.section.transform_name(section)
+        current_value = self.key.get(config, section=section)
+        if current_value == self.value:
+            return ""
+        return f"[{section}] {self.key.key}={current_value}\n"
+
+
+class ConfigDescriptionSection:
```
Comment (SvizelPritula) [inline pisek/config/did_you_mean.py:50]: Shouldn't this be called `ConfigSectionDescription`?

File: pisek/config/did_you_mean.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        if self.similarity_function is None:
+            return 5 * editdistance.distance(self.section, section)
+        else:
+            return self.similarity_function(self.section, section)
+
+    def transform_name(self, name: str) -> str:
+        return name if self.similarity(name) == 0 else self.section
+
+
+class ConfigDescriptionKey:
```
Comment (SvizelPritula) [inline pisek/config/did_you_mean.py:66]: Shouldn't this be called `ConfigKeyDescription`?

File: pisek/config/did_you_mean.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                    section = ConfigDescriptionSection(
+                        line.removeprefix("[").removesuffix("]")
+                    )
+                    self.sections[section.section] = section
+
+                else:
+                    raise ValueError(f"invalid config-description line: '{line}'")
+
+        for key, lambda_cond in add_applicability_conditions:
+            key.applicability_conditions.append(lambda_cond())
```
Comment (SvizelPritula) [inline pisek/config/did_you_mean.py:158]: Why is this deferred until the end of parsing?

File: pisek/config/did_you_mean.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        return self.section.similarity(section) + self.similarity(key)
+
+    def applicable(self, section: str, config: "ConfigHierarchy") -> Optional[str]:
+        text = ""
+        for cond in self.applicability_conditions:
+            text += cond.check(section, config)
+        return text
+
+    def guess(self, section: str, key: str) -> tuple[str, str]:
+        return (self.section.transform_name(section), self.key)
```
Comment (SvizelPritula) [inline pisek/config/did_you_mean.py:94]: I don't understand what this function is supposed to do.

File: pisek/config/did_you_mean.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    def defaults(self) -> list[tuple[str, str]]:
+        return self.defaults_to + [(d, self.key) for d in self.section.defaults_to]
+
+    def similarity(self, key: str) -> int:
+        return editdistance.distance(self.key, key)
+
+    def score(self, section: str, key: str) -> int:
+        return self.section.similarity(section) + self.similarity(key)
+
+    def applicable(self, section: str, config: "ConfigHierarchy") -> Optional[str]:
```
Comment (SvizelPritula) [inline pisek/config/did_you_mean.py:87]: This will always return a `str`.

File: pisek/config/did_you_mean.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    def _invalid_function_args(self, fun_name: str, args: list[str]) -> None:
+        raise ValueError(
+            f"invalid config-description function {fun_name} arguments: '{' '.join(args)}'"
+        )
+
+    def find_section(self, section: str) -> ConfigDescriptionSection:
+        return min(
+            self.sections.values(),
+            key=lambda s: editdistance.distance(section, s.section),
+        )
```
Comment (SvizelPritula) [inline pisek/config/did_you_mean.py:177]: Is this function used anywhere?

**Review by SvizelPritula — CHANGES_REQUESTED** [review]
There are a couple of issues.

File: pisek/config/task_config.py (hunk 1 above)
Comment (5K1PY) [reply pisek/config/task_config.py:243]: Not yet

File: pisek/config/did_you_mean.py (hunk 5 above)
Comment (5K1PY) [reply pisek/config/did_you_mean.py:158]: Because this key might not have been loaded yet

File: pisek/config/config-description
Comment (SvizelPritula) [inline pisek/config/config-description:1]: This file is not part of the package installation and therefore unavailable.

**Review by SvizelPritula — APPROVED** [review]
LGTM

**Pull request merged**